db/limitersMgr.go

```diff
@@ -0,0 +1,696 @@
+package db
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/distributedio/titan/conf"
+	"github.com/distributedio/titan/metrics"
+	sdk_kv "github.com/pingcap/tidb/kv"
+	"go.uber.org/zap"
+	"golang.org/x/time/rate"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	LIMITDATA_DBID             = 0
+	ALL_NAMESPACE              = "*"
+	NAMESPACE_COMMAND_TOKEN    = "@"
+	QPS_PREFIX                 = "qps:"
+	RATE_PREFIX                = "rate:"
+	LIMIT_VALUE_TOKEN          = " "
+	LIMITER_STATUS_PREFIX      = "limiter_status:"
+	LIMITER_STATUS_VALUE_TOKEN = ","
+	TIME_FORMAT                = "2006-01-02 15:04:05"
+	MAXIMUM_WEIGHT             = 1
+	MINIMUM_WEIGHT             = 0.1
+)
+
+type LimiterWrapper struct {
+	limiterName  string
+	limiter      *rate.Limiter
+	globalLimit  int64
+	localPercent float64
+	lock         sync.Mutex
+}
+
+type CommandLimiter struct {
+	localIp     string
+	limiterName string
+
+	qpsLw  LimiterWrapper
+	rateLw LimiterWrapper
+	weight float64
+
+	lock               sync.Mutex
+	skipBalance        bool
+	lastTime           time.Time
+	totalCommandsCount int64
+	totalCommandsSize  int64
+}
+
+type LimitData struct {
+	limit int64
+	burst int
+}
+
+type LimitersMgr struct {
+	limitDatadb *DB
+	conf        *conf.RateLimit
+	localIp     string
+
+	limiters          sync.Map
+	qpsAllmatchLimit  sync.Map
+	rateAllmatchLimit sync.Map
+	lock              sync.Mutex
+}
+
+func getAllmatchLimiterName(limiterName string) string {
+	strs := strings.Split(limiterName, NAMESPACE_COMMAND_TOKEN)
+	if len(strs) < 2 {
+		return ""
+	}
+	return fmt.Sprintf("%s%s%s", ALL_NAMESPACE, NAMESPACE_COMMAND_TOKEN, strs[1])
+}
+
+func getLimiterKey(limiterName string) []byte {
+	var key []byte
+	key = append(key, []byte(LIMITER_STATUS_PREFIX)...)
+	key = append(key, []byte(limiterName)...)
+	key = append(key, ':')
+	return key
+}
+
+func getNamespaceAndCmd(limiterName string) []string {
+	strs := strings.Split(limiterName, NAMESPACE_COMMAND_TOKEN)
+	if len(strs) < 2 {
+		return nil
+	}
+	return strs
+
+}
+
+func NewLimitersMgr(store *RedisStore, rateLimit *conf.RateLimit) (*LimitersMgr, error) {
+	var addrs []net.Addr
+	var err error
+	if rateLimit.InterfaceName != "" {
+		iface, err := net.InterfaceByName(rateLimit.InterfaceName)
+		if err != nil {
+			return nil, err
+		}
+
+		addrs, err = iface.Addrs()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		addrs, err = net.InterfaceAddrs()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	localIp := ""
+	for _, a := range addrs {
+		if ipnet, ok := a.(*net.IPNet); ok && !ipnet.IP.IsLoopback() && ipnet.IP.To4() != nil {
+			localIp = ipnet.IP.String()
+			break
+		}
+	}
+	if localIp == "" {
+		return nil, errors.New(rateLimit.InterfaceName + " adds is empty")
+	}
+
+	if rateLimit.LimiterNamespace == "" {
```

Comment:
If there is no namespace supplied, can we just regard it as a global limitation?
@piaoairy219